src/QirRuntime/lib/QIR/callables.cpp

```diff
@@ -101,19 +139,34 @@ extern "C"
 ==============================================================================*/
 int QirTupleHeader::AddRef()
 {
-    assert(refCount > 0);
-    return ++refCount;
+    assert(this->refCount > 0);
+    return ++this->refCount;
 }
 
 int QirTupleHeader::Release()
 {
-    --refCount;
-    if (refCount == 0)
+    assert(this->refCount > 0); // doesn't guarantee we catch double releases but better than nothing
+    --this->refCount;
+    if (this->refCount == 0)
     {
         char* buffer = reinterpret_cast<char*>(this);
         delete[] buffer;
     }
-    return refCount;
+    return this->refCount;
+}
+
+void QirTupleHeader::AddUser()
+{
+    ++this->userCount;
+}
+
+void QirTupleHeader::RemoveUser()
+{
+    if (this->userCount == 0)

```

Comment:
This may already be covered in the spec somewhere, but is the assumption that this QirRuntime is not natively thread safe? Since there is no locking or other enforced atomicity here, this check doesn't guarantee that a tuple user count won't be negative if multiple threads access the same object (same for arrays above).

Reply:
Ah, I see now where we assert that the existing runtime is assumed to be single threaded. Makes sense!